service/service.go

```diff
@@ -212,40 +231,55 @@ func (s *Service) Start(b Broadcaster) error {
 	} else if err := s.Recover(); err != nil {
 		return fmt.Errorf("failed to recover: %v", err)
 	}
-
+	now := time.Now()
+	epoch := time.Duration(0)
+	if d := now.Sub(s.genesis); d > 0 {
+		epoch = d / s.cfg.EpochDuration
+	}
 	if s.openRound == nil {
-		s.openRound = s.newRound()
+		s.openRound = s.newRound(uint32(epoch))
 		log.Info("Round %v opened", s.openRound.ID)
 	}
 
 	go func() {
+		timer := time.NewTimer(0)
+		<-timer.C
+		defer timer.Stop()
 		for {
-			select {
-			case <-s.openRoundClosure():
-			case <-s.sig.ShutdownRequestedChan:
-				log.Info("Shutdown requested, service shutting down")
-				s.openRound = nil
-				return
+			start := s.genesis.
+				Add(s.cfg.EpochDuration * time.Duration(s.openRound.Epoch())).
+				Add(s.cfg.PhaseShift)
+			if d := start.Sub(time.Now()); d > 0 {
+				log.Info("Round %v waiting for execution to start for %v",
+					s.openRound.ID, d)
+				timer.Reset(d)
+				select {
+				case <-timer.C:
+				case <-s.sig.ShutdownRequestedChan:
+					log.Info("Shutdown requested, service shutting down")
+					s.openRound = nil
+					return
+				}
 			}
 
 			if s.openRound.isEmpty() && !s.cfg.ExecuteEmpty {
 				continue
 			}
 
 			s.prevRound = s.openRound
-			s.openRound = s.newRound()
-			log.Info("Round %v opened", s.openRound.ID)
+			open := time.Now()
+			// TODO(dshulyak) some time is wasted here to persist data on disk
+			// on my computer ~20ms
```

Comment:
wasted how? you mean it delays the broadcasting by 20ms? maybe this comment should be on line 281?

Reply:
so each round has a deadline at next_round_start - cycle_gap. ideally we would run poet without wasting time till this deadline. but we are wasting some time because of the inefficiency that i described in this todo